### ISSUE TITLE
SpriteBatch: Any changes 3.2 to 3.3?

### ISSUE THREAD

**Issue by paewie** [mono/MonoGame#3618]
As I tried to update to MonoGame 3.3 I encountered some strange behavior. I'm really unsure if this is a bug or me misusing something, but since it worked fine under 3.2 and fails in 3.3 now... My guess would be some different handling within the SpriteBatch class, but I have no idea of the actual Framework code at all.

First of all: I'm using the "Advanced Zima 2D Shadows" (MonoGame version), compiled with the 3.3 Pipeline Tool on Windows and using the 3.2 assemblies of MonoGame (in my actual game) everything just worked fine:
![tm_shader_working](https://cloud.githubusercontent.com/assets/11522235/6690867/5137444e-ccc3-11e4-928f-1425daa8a052.png)

Now I switched to the 3.3 assemblies and got this result (without any other changes):
![tm_shader_failing](https://cloud.githubusercontent.com/assets/11522235/6690878/654d230e-ccc3-11e4-9839-6db971301e4d.png)

After some research I could narrow down the Problem to a 2x multi blend shader that doesn't seem to do it's job anymore.
![tm_shader_failinginfo](https://cloud.githubusercontent.com/assets/11522235/6690899/7ec7a426-ccc3-11e4-9122-095c85eb1a25.png)
As you can see the lightmap is still being created correctly, but the blending part just blackens everything out (I used my main SpriteBatch to simply draw the lightmap instead of blending it with the underlying texture as done in the first picture). In the actual code there is a separate SpriteBatch set to use the multi blend shader to draw both textures.
Note: The rocks around the player/placeholder changed since I restarted/rebuilt the game each time and therefore the map was randomly rebuilt each time.

Since I'm quite confused now whether it's my bad or not I couldn't help myself besides posting this as an issue here. I am working on a Linux platform and even tried the WinGL assembly, but I still got this (unwanted) result.

The c# code where this issue occurs looks like this (completly from the shading system mentioned above):

```
this.LightBlender.Parameters["MixFactor"].SetValue(mixFactor0to1);
graphics.GraphicsDevice.SetRenderTarget(renderTarget);
spriteBatch.Begin(SpriteSortMode.Immediate, BlendState.AlphaBlend, SamplerState.LinearClamp, null, RasterizerState.CullNone, this.LightBlender);
graphics.GraphicsDevice.Textures[1] = light;
graphics.GraphicsDevice.Textures[1].GraphicsDevice.SamplerStates[1] = SamplerState.LinearClamp;
spriteBatch.Draw(underlyingTexture, Vector2.Zero, Color.White);
spriteBatch.End();
```

Sorry, if this is my issue and not yours, but I don't quite understand shaders yet and I didn't make any changes except the assembly-version I use (both are marked as 3.1.2.0 though).
Oh, while writing this I remembered: The "old" assembly was not the official 3.2 one. It was the one from the "MonoDevelop Alpha Add-In Package" (.mpack) that @cra0zy (or however his name is spelled, sorry) posted. I'm unsure which version this actually is since they are all marked as 3.1.2.0...

**Comment by tomspilman** [comment]
>  I'm really unsure if this is a bug or me misusing something, but since it worked fine under 3.2 and fails in 3.3 now

It really could be either as many things have been changed, so lets investigate.

So first thing... what does this "LightBlender" effect look like?

**Comment by paewie** [comment]
Oops, you're right. I forgot that code :D

sampler TextureSampler : register(s0);
sampler LightSampler : register(s1);

float MixFactor = 1;
float4 Portion = 0;
float PortionScale = 1;

float4 MultiBlend(float2 texCoord : TEXCOORD0, float4 color : COLOR0) : COLOR0
{
  float4 colorGround = tex2D(TextureSampler, texCoord);
  float4 colorShadow = 0;

   if (Portion[2] == 0 || Portion[3] == 0)
  {
	colorShadow = tex2D(LightSampler, texCoord);
  }
  else
  {
	float2 newCoord = float2(Portion.x + (Portion.z * texCoord.x),Portion.y + (Portion.w * texCoord.y)) * PortionScale;
	colorShadow = tex2D(LightSampler, newCoord);
  }

  float4 resultColor = colorGround * colorShadow + colorShadow * colorGround;

  return lerp(colorGround, resultColor, MixFactor);
}

technique Basic
{
	pass Pass0
	{
		PixelShader = compile ps_2_0 MultiBlend();
	}
}

**Comment by tomspilman** [comment]
You said this is running on Linux?  I ask because there are additional rules for shaders on Windows and DirectX.

**Comment by paewie** [comment]
Yes, I'm on Linux Mint 17.1 right now. If it could make any difference I could try to use Win7 x64, too.

**Comment by cra0zy** [comment]
You got my name right :D,  anyway, I provided development builds, not stable 3.2 builds.

**Comment by paewie** [comment]
Ah okay. I alredy got that feeling while investigating that issue myself, but good to know :)

**Comment by paewie** [comment]
Originally this shader was written for Windows and DX platforms, though. I got it from http://www.funhazard.com/xna-resources.html since I thought it looks quite nice and is free to use. There is a special MonoGame version which I'm already using.

**Comment by tomspilman** [comment]
First thing to check.  The default values used in:
```
float MixFactor = 1;
float4 Portion = 0;
float PortionScale = 1;
```
Are not supported under GLSL... we don't get the defaults from MojoShader.  So you need to be setting these from the C# side if not they will default to 0.

**Comment by paewie** [comment]
Well, the MixFactor is set to 0.95f in this case. Setting it to 1 results in an black background image and to 0 in just the background image (without any brightness changes). Manually setting the PortionScale to 1 doesn't seem to have any effect and neither has the (seemingly redundant) setting of Portion to 0...
Even if this shows my small knowledge about shaders, it sadly doesn't have any effect on the result...

**Comment by paewie** [comment]
Okay, as far as i understood the shader it does nothing besides calculating a result color and returning the lineary interpolated value between the ground's color and the result color relative to the mix factor.
This makes me rather sceptical that the "error" originates from the shader. If it hadn't worked in this earlier development build of MonoGame I would suspect my usage of the shader (which is actually an almost exact copy of the example usage). But this way I'm just confused and kind of helpless. The most complex parts of the shader work but the simplest doesn't. The only difference I can figure out is the usage within the "SpriteBatch.Begin" method, as any other shader code is executed "directly" via "CurrentTechnique.Passes[0].Apply()" and customly rendering the output to a texture/RenderTarget2D. I could try to copy (and modify) that custom rendering to apply the blending, but I'm quite sure it will overcome my knowledge of how to access shaders from c#...

That would be my only idea on how to solve this problem since I don't know anything about the internal things of a SpriteBatch...
Either this or a somehow mistaken (?) association between those registers (shader lines 1 and 2) and the GraphicsDevice's texture collection.

**Comment by Shqrdx** [comment]
>sampler TextureSampler : register(s0);
>sampler LightSampler : register(s1);

I checked the internal SpriteEffect and found it used same register for sprite texture. Try remove registers from this effect.

**Comment by tomspilman** [comment]
> Try remove registers from this effect.

I don't think that will help.  In fact I think the registers are needed to make it work at all.

**Comment by paewie** [comment]
Well, actually it worked under Linux before (with the older assembly) and since I use the official NVIDIA proprietary driver there shouldn't be a problem with it.
Anyways, I already thought of replacing this code, too. I'm too unexperienced to know, so are there any other possibilities to pass the textures on to the shader (like setting it as an parameter like the MixFactor)?

**Comment by paewie** [comment]
Just to report what I found out:
I dug through the code of a SpriteBatch and weren't able to find anything that could affect the shader or the TextureCollection. I don't know where the rendering or more precisely shader "execution" goes on from there, but (besides the output) it doesn't seem to do anything unexpected... In fact the other shaders work as they did before (see third image). The textures are still being correctly assigned to the TextureCollection (at least the debugger sees they are there), the shader still draws the underlying texture lineary interpolated using the MixFactor but instead of using the light texture it uses a completely black one.
So I haven't got any clue what changed / stopped functioning or how to solve this situation except leaving this shadow system out for now. I'm really sorry I'm that less helpful, but I'm having my exams right now and are going to start with a part time job in the near future - in short not enough time to examine a "foreign" assembly that thoroughly. Hopefully you guys can find anything that's not working :)

**Comment by tgjones** [comment]
@paewie Yes, there was a change in `SpriteBatch` behaviour between 3.2 and 3.3. Instead of this:

``` csharp
graphics.GraphicsDevice.Textures[1] = light;
```

try doing this:

``` csharp
this.LightBlender.Parameters["LightSampler"].SetValue(light);
```

The behaviour was changed to match XNA: effect parameter values are now prioritised over whatever is in `GraphicsDevice.Textures`.

**Comment by paewie** [comment]
@tgjones this did the trick. You can't believe how happy I am. You are my savior of the week :)
Thanks.

**Comment by tomspilman** [comment]
@tgjones - So is this a case where his code was invalid even for XNA?  Or is this an incompatibility in MonoGame?

**Comment by tgjones** [comment]
@tomspilman To the best of my knowledge (based on a number of tests against XNA), there *was* an incompatibility in MonoGame 3.2, which was fixed in 3.3. MonoGame now matches XNA behaviour.

In 3.2, MonoGame didn't apply textures in effect parameters if they were null. This didn't match XNA, which always applied them. So in 3.3, MonoGame sets whatever is in texture effect parameter values into GraphicsDevice.Textures, even if it's null. This is what #3554 was about.

So yes, I believe his code was invalid even for XNA.

**Comment by tomspilman** [comment]
Calling this issue closed then!  Thanks @paewie and @tgjones !
